### PR TITLE
49 implement post processing

### LINKED ISSUE #49: Implement Post Processing
# Goal

Write an implementation for the already written post processing interface

### PR DESCRIPTION
**Pull request by miklasr**
# Details

- Add implementation for post processing
- Add corresponding tests
- Rework the testing setup
- Refactor internal data structures to remove repetitive code
    - Make certain areas of code less error-prone

# Checklist

## Code
- [x] I have added some tests
- [x] I have run the tests locally to ensure they all pass by running `./test.sh`

## Documentation
- [x] I have updated documentation appropriately (GitHub Pages)
- [x] I have tested the updated GitHub Pages locally to ensure the changes to documentation render correctly (see this [Guidance](https://ese-peasy.github.io/PosturePerfection/guidance.html) page)

Closes #49

### CODE REVIEW

**Comment by miklasr** [comment]
**Note**: The current `mian()` doesn't use my post processor, but the module has been tested with unit testing. I think we need to include multi-frame inputs, before the post-processor can be properly integration tested.

File: src/iir.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+float IIR2ndOrderFilter::run(float x) {
+  float output = this->nodes.b1 * this->nodes.tap1;
+  x = x - (this->nodes.a1 * this->nodes.tap1);
+  output = output + (this->nodes.b2 * this->nodes.tap2);
+  x = x - (this->nodes.a2 * this->nodes.tap2);
+  output = output + x * this->nodes.b0;
+  this->nodes.tap2 = this->nodes.tap1;
+  this->nodes.tap1 = x;
+  return output;
+}
```
Comment (Ashwin-MJ) [inline src/iir.cpp:46]: Had to look up my DSP notes for this, looks right to me! Nice stuff 😄

File: src/iir.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+   * @brief Vector of the constituent second-order filters
+   *
+   */
+  std::vector<IIR2ndOrderFilter> filters;
+
+ public:
+  /**
+   * @brief Construct a new `IIRFilter` object
+   *
+   * Passing empty `SmoothingSettings` disbales the IIR filter.
```
Comment (Ashwin-MJ) [inline src/iir.h]: *Minor*: ~disbales~ -> disables

File: src/intermediate_structures.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (Ashwin-MJ) [inline src/intermediate_structures.h:1]: Big fan of this change, makes the code **so** much nicer! 👍 

Only thing is perhaps adding a bit of documentation to the undocumented structs/arrays. I don't mind accepting this PR and then adding this in myself, but I think there should be some explanation why we have two `Coordinate` structs)

File: src/post_processor.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                       : Status::Untrustworthy};
+    body_part_index++;
+  }
+
+  body_part_index = BodyPartMin;
+  for (auto& body_part : results.body_parts) {
+    body_part.x = this->iir_filters.at(body_part_index).run(body_part.x);
+    body_part.y = this->iir_filters.at(body_part_index + 1).run(body_part.y);
+    body_part_index += 2;
+  }
```
Comment (Ashwin-MJ) [inline src/post_processor.cpp]: Not sure if I'm being silly, but I don't entirely understand how this works?

File: src/iir.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ */
+#ifndef SRC_IIR_H_
+#define SRC_IIR_H_
+
+#include <stdlib.h>
+
+#include <vector>
+
+#define COEFFS_PER_IIR2ndOrderFilter 6  //  Coefficients per filter stage
```
Comment (Ashwin-MJ) [inline src/iir.h]: Do we use this anywhere?

**Review by Ashwin-MJ — COMMENTED** [review]
Awesome work, all looks really great! Moving to `intermediate_structures.h` should make things much better to work with. The tests look great too!

A few very minor comments, mainly for my own understanding.

I will run this on my machine too and let you know how it goes

File: src/intermediate_structures.h (hunk 3 above)
Comment (Ashwin-MJ) [reply src/intermediate_structures.h:1]: (This was meant to be a comment for the whole file)

File: src/post_processor.cpp (hunk 4 above)
Comment (miklasr) [reply src/post_processor.cpp]: I'll comment this to make it clearer

**Comment by Ashwin-MJ** [comment]
Built and passed on my machine using `./test.sh` 🎉

File: src/post_processor.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                       ? Status::Trustworthy
+                       : Status::Untrustworthy};
+    body_part_index++;
+  }
+
+  body_part_index = BodyPartMin;
+  for (auto& body_part : results.body_parts) {
+    body_part.x = this->iir_filters.at(body_part_index).run(body_part.x);
+    body_part.y = this->iir_filters.at(body_part_index + 1).run(body_part.y);
+    body_part_index += 2;
```
Comment (miklasr) [inline src/post_processor.cpp]: The filtering should be done **before** the `Trustworthy`/`Untrustworthy` label is applied. I'll fix this.

File: src/iir.h (hunk 2 above)
Comment (miklasr) [reply src/iir.h]: Fixed

File: src/intermediate_structures.h (hunk 3 above)
Comment (miklasr) [reply src/intermediate_structures.h:1]: Thanks, I've updated the documentation and tried to make the distinction between the coordinate types clearer.

File: src/post_processor.cpp (hunk 4 above)
Comment (miklasr) [reply src/post_processor.cpp]: The loop has been merged with the other one in the file. I've also added comments that hopefully make it clearer. Let me know if you think I should make it even clearer or make the use of multiple filters more explicit in some other way.

File: src/iir.h (hunk 5 above)
Comment (miklasr) [reply src/iir.h]: Not used; removed

File: src/post_processor.cpp (hunk 6 above)
Comment (miklasr) [reply src/post_processor.cpp]: I've combined the loops and made sure filtering is done first now. At the moment it doesn't matter, but if we do try out filtering confidence values in time (not sure yet if that would make sense) the change is now trivial.

**Comment by miklasr** [comment]
I've made some changes following the comments and I think it should definitely be good to merge now.

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -2,6 +2,14 @@ set(CMAKE_CXX_STANDARD 11)
 cmake_minimum_required(VERSION 3.16) # Required by TFL
 project(PosturePerfection VERSION 0.1.0)
 
+
+set(LIBSRC
+  doubler.cpp
```
Comment (C-Begley) [inline src/CMakeLists.txt]: Should we give this a more meaningful name or even just remove this?

**Review by Ashwin-MJ — APPROVED** [review]
Changes look good to me!

File: src/CMakeLists.txt (hunk 7 above)
Comment (miklasr) [reply src/CMakeLists.txt]: Doubler has been removed. I forgot to do that so thanks for spotting!

**Review by Ashwin-MJ — APPROVED** [review]

**Pull request merged**